File: android/app/src/main/java/tech/loveace/appv3/data/service/ISIMService.kt
package tech.loveace.appv3.data.service

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.Headers
import org.jsoup.Jsoup
import tech.loveace.appv3.data.model.*
import tech.loveace.appv3.data.network.AUFEConnection
import tech.loveace.appv3.data.network.HttpClient

/**
 * ISIM 宿舍电费服务
 */
class ISIMService(private val connection: AUFEConnection) {

    private val isimClient = HttpClient(baseUrl = BASE_URL, timeoutMs = 30_000, followRedirects = true)
    private var jsessionId: String? = null
    private var sessionInitialized = false

    private suspend fun ensureSession() {
        if (sessionInitialized && jsessionId != null) return
        withContext(Dispatchers.IO) {
            isimClient.copyCookiesFrom(connection.client)
            val url = "$BASE_URL/go?openid=${connection.userId}&sn=sn"
            isimClient.get(url)
            jsessionId = isimClient.cookieJar.getCookie("JSESSIONID")
            if (jsessionId != null) sessionInitialized = true
            else throw Exception("无法获取 JSESSIONID")
        }
    }

    suspend fun getBuildings(): UniResponse<List<Map<String, String>>> = withContext(Dispatchers.IO) {
        try {
            ensureSession()
            val twfId = connection.twfId ?: ""
            val response = isimClient.get("$BASE_URL/about", headers = mapOf(
                "Cookie" to "JSESSIONID=$jsessionId; TWFID=$twfId",
            ))
            val html = response.body?.string() ?: throw Exception("响应为空")
            val doc = Jsoup.parse(html)
            val buildings = mutableListOf<Map<String, String>>()
            for (script in doc.select("script")) {
                val content = script.data()
                if (content.contains("pickerBuilding")) {
                    val valuesMatch = Regex("values:\\s*\\[(.*?)]").find(content)
                    val displayMatch = Regex("displayValues:\\s*\\[(.*?)]").find(content)
                    if (valuesMatch != null && displayMatch != null) {
                        val values = valuesMatch.groupValues[1].split(",").map { it.trim().replace("\"", "").replace("'", "") }.filter { it.isNotEmpty() }
                        val displays = displayMatch.groupValues[1].split(",").map { it.trim().replace("\"", "").replace("'", "") }.filter { it.isNotEmpty() && it != "请选择" }
                        for (i in 0 until minOf(values.size, displays.size)) {
                            if (values[i].isNotEmpty()) buildings.add(mapOf("code" to values[i], "name" to displays[i]))
                        }
                    }
                    break
                }
            }
            UniResponse.success(buildings)
        } catch (e: Exception) {
            Log.e(TAG, "getBuildings failed", e)
            UniResponse.failure(e.message ?: "获取楼栋列表失败")
        }
    }

    suspend fun getFloors(buildingCode: String): UniResponse<List<Map<String, String>>> = withContext(Dispatchers.IO) {
        try {
            ensureSession()
            val twfId = connection.twfId ?: ""
            val response = isimClient.get("$BASE_URL/about/floors/$buildingCode", headers = mapOf(
                "Cookie" to "JSESSIONID=$jsessionId; TWFID=$twfId",
                "X-Requested-With" to "XMLHttpRequest",
            ))
            val body = response.body?.string() ?: "[]"
            val parsed = parseFloorRoomJson(body, "floordm", "floorname")
            UniResponse.success(parsed)
        } catch (e: Exception) {
            Log.e(TAG, "getFloors failed", e)
            UniResponse.failure(e.message ?: "获取楼层列表失败")
        }
    }

    suspend fun getRooms(floorCode: String): UniResponse<List<Map<String, String>>> = withContext(Dispatchers.IO) {
        try {
            ensureSession()
            val twfId = connection.twfId ?: ""
            val response = isimClient.get("$BASE_URL/about/rooms/$floorCode", headers = mapOf(
                "Cookie" to "JSESSIONID=$jsessionId; TWFID=$twfId",
                "X-Requested-With" to "XMLHttpRequest",
            ))
            val body = response.body?.string() ?: "[]"
            val parsed = parseFloorRoomJson(body, "roomdm", "roomname")
            UniResponse.success(parsed)
        } catch (e: Exception) {
            Log.e(TAG, "getRooms failed", e)
            UniResponse.failure(e.message ?: "获取房间列表失败")
        }
    }

    suspend fun getElectricityInfo(roomCode: String, displayText: String? = null): UniResponse<ElectricityInfo> =
        withContext(Dispatchers.IO) {
            try {
                ensureSession()
                val twfId = connection.twfId ?: ""
                // Bind room
                isimClient.post("$BASE_URL/about/rebinding", formData = mapOf(
                    "roomdm" to roomCode, "room" to (displayText ?: roomCode),
                    "openid" to connection.userId, "sn" to "sn", "mode" to "u",
                ), headers = mapOf("Cookie" to "JSESSIONID=$jsessionId;TWFID=$twfId"))

                val headers = mapOf("Cookie" to "JSESSIONID=$jsessionId;TWFID=$twfId")
                val usageDeferred = async { isimClient.get("$BASE_URL/use/record", headers = headers) }
                val paymentDeferred = async { isimClient.get("$BASE_URL/pay/record", headers = headers) }
                val usageHtml = usageDeferred.await().body?.string() ?: ""
                val paymentHtml = paymentDeferred.await().body?.string() ?: ""

                val balance = parseBalance(usageHtml)
                val usageRecords = parseUsageRecords(usageHtml)
                val paymentRecords = parsePaymentRecords(paymentHtml)
                UniResponse.success(ElectricityInfo(balance, usageRecords, paymentRecords))
            } catch (e: Exception) {
                Log.e(TAG, "getElectricityInfo failed", e)
                UniResponse.failure(e.message ?: "获取电费信息失败", retryable = true)
            }
        }

    private fun parseFloorRoomJson(body: String, codeKey: String, nameKey: String): List<Map<String, String>> {
        val result = mutableListOf<Map<String, String>>()
        try {
            val jsonStr = body.replace(Regex("([a-zA-Z_][a-zA-Z0-9_]*)\\s*:"), "\"$1\":")
            val json = kotlinx.serialization.json.Json.parseToJsonElement(jsonStr)
            val arr = json.jsonArray
            if (arr.isNotEmpty()) {
                val obj = arr[0].jsonObject
                val codes = obj[codeKey]?.jsonArray ?: return result
                val names = obj[nameKey]?.jsonArray ?: return result
                for (i in 1 until minOf(codes.size, names.size)) {
                    val code = codes[i].jsonPrimitive.contentOrNull ?: continue
                    val name = names[i].jsonPrimitive.contentOrNull ?: continue
                    if (code.isNotEmpty() && name != "请选择") result.add(mapOf("code" to code, "name" to name))
                }
            }
        } catch (e: Exception) { Log.w(TAG, "parseFloorRoomJson failed", e) }
        return result
    }

    private fun parseBalance(html: String): ElectricityBalance {
        var purchased = 0.0; var subsidy = 0.0
        val doc = Jsoup.parse(html)
        for (item in doc.select("li.item-content, li.item, .item-content")) {
            val title = item.selectFirst(".item-title, .title, dt, .label")?.text()?.trim() ?: continue
            val value = item.selectFirst(".item-after, .value, dd, .amount")?.text()?.trim() ?: continue
            val amount = Regex("([\\d.]+)").find(value)?.groupValues?.get(1)?.toDoubleOrNull() ?: continue
            when {
                title.contains("购电") -> purchased = amount
                title.contains("补助") -> subsidy = amount
            }
        }
        return ElectricityBalance(purchased, subsidy)
    }

    private fun parseUsageRecords(html: String): List<ElectricityUsageRecord> {
        val records = mutableListOf<ElectricityUsageRecord>()
        val doc = Jsoup.parse(html)
        for (item in doc.select("#divRecord ul li")) {
            val time = item.selectFirst(".item-title")?.text()?.trim() ?: continue
            val usageText = item.selectFirst(".item-after")?.text()?.trim() ?: continue
            val meterText = item.selectFirst(".item-subtitle")?.text()?.trim() ?: ""
            val usage = Regex("([\\d.]+)度").find(usageText)?.groupValues?.get(1)?.toDoubleOrNull() ?: continue
            val meter = Regex("电表:\\s*(.+)").find(meterText)?.groupValues?.get(1)?.trim() ?: meterText
            records.add(ElectricityUsageRecord(time, usage, meter))
        }
        return records
    }

    private fun parsePaymentRecords(html: String): List<PaymentRecord> {
        val records = mutableListOf<PaymentRecord>()
        val doc = Jsoup.parse(html)
        for (item in doc.select("#divRecord ul li")) {
            val time = item.selectFirst(".item-title")?.text()?.trim() ?: continue
            val amountText = item.selectFirst(".item-after")?.text()?.trim() ?: continue
            val typeText = item.selectFirst(".item-subtitle")?.text()?.trim() ?: ""
            val amount = Regex("(-?[\\d.]+)元").find(amountText)?.groupValues?.get(1)?.toDoubleOrNull() ?: continue
            val type = Regex("类型:\\s*(.+)").find(typeText)?.groupValues?.get(1)?.trim() ?: typeText
            records.add(PaymentRecord(time, amount, type))
        }
        return records
    }

    companion object {
        private const val TAG = "ISIMService"
        const val BASE_URL = "http://hqkd-aufe-edu-cn.vpn2.aufe.edu.cn"
    }
}

private val kotlinx.serialization.json.JsonElement.jsonArray get() = this as kotlinx.serialization.json.JsonArray
private val kotlinx.serialization.json.JsonElement.jsonObject get() = this as kotlinx.serialization.json.JsonObject
private val kotlinx.serialization.json.JsonElement.jsonPrimitive get() = this as kotlinx.serialization.json.JsonPrimitive
private val kotlinx.serialization.json.JsonPrimitive.contentOrNull: String? get() = if (isString) content else try { content } catch (_: Exception) { null }

File: ios/loveaceios/Service/ISIMService.swift
import Foundation
import SwiftSoup
import os

private let logger = Logger(subsystem: "tech.loveace.loveaceios", category: "ISIMService")

actor ISIMService {
    private let connection: AUFEConnection
    private var isimClient: HTTPClient?
    private var jsessionId: String?
    private var sessionInitialized = false
    static let baseURL = "http://hqkd-aufe-edu-cn.vpn2.aufe.edu.cn"

    init(connection: AUFEConnection) {
        self.connection = connection
    }

    private func ensureSession() async throws {
        if sessionInitialized && jsessionId != nil { return }
        let mainClient = await connection.client!
        let client = HTTPClient(baseUrl: Self.baseURL, timeoutInterval: 90, cookieStorage: mainClient.cookieStorage)
        self.isimClient = client
        let userId = await connection.userId
        let url = "\(Self.baseURL)/go?openid=\(userId)&sn=sn"
        let (_, response) = try await client.get(url)
        if let respUrl = response.url?.absoluteString,
           let range = respUrl.range(of: "jsessionid=([^?&;]+)", options: [.regularExpression, .caseInsensitive]) {
            jsessionId = String(respUrl[range]).replacingOccurrences(of: "jsessionid=", with: "", options: .caseInsensitive)
        }
        logger.info("ISIM session: jsessionId=\(self.jsessionId ?? "nil")")
        guard jsessionId != nil else { throw ServiceError.parseError("无法获取 JSESSIONID") }
        sessionInitialized = true
    }

    private func sessionHeaders() async -> [String: String] {
        let twfId = await connection.twfId ?? ""
        var cookie = "TWFID=\(twfId)"
        if let jid = jsessionId { cookie = "JSESSIONID=\(jid); \(cookie)" }
        return ["Cookie": cookie]
    }

    func getBuildings() async -> UniResponse<[[String: String]]> {
        do {
            try await ensureSession()
            guard let client = isimClient else { throw ServiceError.emptyResponse }
            let headers = await sessionHeaders()
            let (data, _) = try await client.get("\(Self.baseURL)/about", headers: headers)
            let html = String(data: data, encoding: .utf8) ?? ""
            let doc = try SwiftSoup.parse(html)
            var buildings: [[String: String]] = []
            for script in try doc.select("script").array() {
                let content = try script.data()
                guard content.contains("pickerBuilding") else { continue }
                if let valuesMatch = content.range(of: "values:\\s*\\[(.*?)\\]", options: .regularExpression),
                   let displayMatch = content.range(of: "displayValues:\\s*\\[(.*?)\\]", options: .regularExpression) {
                    let valuesStr = String(content[valuesMatch]).replacingOccurrences(of: "values:", with: "")
                        .trimmingCharacters(in: CharacterSet(charactersIn: " []"))
                    let displaysStr = String(content[displayMatch]).replacingOccurrences(of: "displayValues:", with: "")
                        .trimmingCharacters(in: CharacterSet(charactersIn: " []"))
                    let values = valuesStr.split(separator: ",").map { $0.trimmingCharacters(in: .whitespaces).replacingOccurrences(of: "\"", with: "").replacingOccurrences(of: "'", with: "") }
                    let displays = displaysStr.split(separator: ",").map { $0.trimmingCharacters(in: .whitespaces).replacingOccurrences(of: "\"", with: "").replacingOccurrences(of: "'", with: "") }
                    for i in 0..<min(values.count, displays.count) {
                        let code = values[i], name = displays[i]
                        if !code.isEmpty && !name.isEmpty && name != "请选择" {
                            buildings.append(["code": code, "name": name])
                        }
                    }
                }
                break
            }
            return .success(buildings)
        } catch {
            logger.error("getBuildings failed: \(error.localizedDescription)")
            return .failure(error.localizedDescription)
        }
    }

    func getFloors(buildingCode: String) async -> UniResponse<[[String: String]]> {
        do {
            try await ensureSession()
            guard let client = isimClient else { throw ServiceError.emptyResponse }
            var headers = await sessionHeaders()
            headers["X-Requested-With"] = "XMLHttpRequest"
            let (data, _) = try await client.get("\(Self.baseURL)/about/floors/\(buildingCode)", headers: headers)
            let body = String(data: data, encoding: .utf8) ?? "[]"
            logger.info("getFloors(\(buildingCode)): body=\(body.prefix(300))")
            return .success(parseFloorRoomJSON(body, codeKey: "floordm", nameKey: "floorname"))
        } catch {
            logger.error("getFloors failed: \(error.localizedDescription)")
            return .failure(error.localizedDescription)
        }
    }

    func getRooms(floorCode: String) async -> UniResponse<[[String: String]]> {
        do {
            try await ensureSession()
            guard let client = isimClient else { throw ServiceError.emptyResponse }
            var headers = await sessionHeaders()
            headers["X-Requested-With"] = "XMLHttpRequest"
            let (data, _) = try await client.get("\(Self.baseURL)/about/rooms/\(floorCode)", headers: headers)
            let body = String(data: data, encoding: .utf8) ?? "[]"
            return .success(parseFloorRoomJSON(body, codeKey: "roomdm", nameKey: "roomname"))
        } catch {
            logger.error("getRooms failed: \(error.localizedDescription)")
            return .failure(error.localizedDescription)
        }
    }

    func getElectricityInfo(roomCode: String, displayText: String? = nil) async -> UniResponse<ElectricityInfo> {
        do {
            try await ensureSession()
            guard let client = isimClient else { throw ServiceError.emptyResponse }
            let userId = await connection.userId
            let headers = await sessionHeaders()
            _ = try await client.post("\(Self.baseURL)/about/rebinding", formData: [
                "roomdm": roomCode, "room": displayText ?? roomCode,
                "openid": userId, "sn": "sn", "mode": "u"
            ], headers: headers)

            async let usageResult = client.get("\(Self.baseURL)/use/record", headers: headers)
            async let paymentResult = client.get("\(Self.baseURL)/pay/record", headers: headers)
            let (usageData, _) = try await usageResult
            let (paymentData, _) = try await paymentResult
            let usageHtml = String(data: usageData, encoding: .utf8) ?? ""
            let paymentHtml = String(data: paymentData, encoding: .utf8) ?? ""

            let balance = parseBalance(usageHtml)
            let usageRecords = parseUsageRecords(usageHtml)
            let paymentRecords = parsePaymentRecords(paymentHtml)
            return .success(ElectricityInfo(balance: balance, usageRecords: usageRecords, payments: paymentRecords))
        } catch {
            logger.error("getElectricityInfo failed: \(error.localizedDescription)")
            return .failure(error.localizedDescription, retryable: true)
        }
    }

    private func parseFloorRoomJSON(_ body: String, codeKey: String, nameKey: String) -> [[String: String]] {
        var result: [[String: String]] = []
        let normalized = body.replacingOccurrences(of: "([a-zA-Z_][a-zA-Z0-9_]*)\\s*:", with: "\"$1\":", options: .regularExpression)
        guard let data = normalized.data(using: .utf8),
              let arr = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]],
              let obj = arr.first,
              let codes = obj[codeKey] as? [Any],
              let names = obj[nameKey] as? [Any] else { return result }
        for i in 1..<min(codes.count, names.count) {
            let code = "\(codes[i])"
            let name = "\(names[i])"
            guard !code.isEmpty, name != "请选择" else { continue }
            result.append(["code": code, "name": name])
        }
        return result
    }

    private func parseBalance(_ html: String) -> ElectricityBalance {
        var purchased = 0.0, subsidy = 0.0
        guard let doc = try? SwiftSoup.parse(html) else { return ElectricityBalance() }
        let selectors = "li.item-content, li.item, .item-content"
        guard let items = try? doc.select(selectors) else { return ElectricityBalance() }
        for item in items {
            guard let title = try? item.select(".item-title, .title, dt, .label").first()?.text().trimmingCharacters(in: .whitespaces),
                  let value = try? item.select(".item-after, .value, dd, .amount").first()?.text().trimmingCharacters(in: .whitespaces) else { continue }
            guard let amountMatch = value.range(of: "([\\d.]+)", options: .regularExpression),
                  let amount = Double(value[amountMatch]) else { continue }
            if title.contains("购电") { purchased = amount }
            else if title.contains("补助") { subsidy = amount }
        }
        return ElectricityBalance(remainingPurchased: purchased, remainingSubsidy: subsidy)
    }

    private func parseUsageRecords(_ html: String) -> [ElectricityUsageRecord] {
        guard let doc = try? SwiftSoup.parse(html),
              let items = try? doc.select("#divRecord ul li") else { return [] }
        return items.compactMap { item in
            guard let time = try? item.select(".item-title").first()?.text().trimmingCharacters(in: .whitespaces),
                  let usageText = try? item.select(".item-after").first()?.text().trimmingCharacters(in: .whitespaces),
                  let usageMatch = usageText.range(of: "([\\d.]+)度", options: .regularExpression) else { return nil }
            let usageStr = usageText[usageMatch].replacingOccurrences(of: "度", with: "")
            guard let usage = Double(usageStr) else { return nil }
            let meterText = (try? item.select(".item-subtitle").first()?.text().trimmingCharacters(in: .whitespaces)) ?? ""
            let meter = meterText.range(of: "电表:\\s*(.+)", options: .regularExpression).map { String(meterText[$0]).replacingOccurrences(of: "电表:", with: "").trimmingCharacters(in: .whitespaces) } ?? meterText
            return ElectricityUsageRecord(recordTime: time, usageAmount: usage, meterName: meter)
        }
    }

    private func parsePaymentRecords(_ html: String) -> [PaymentRecord] {
        guard let doc = try? SwiftSoup.parse(html),
              let items = try? doc.select("#divRecord ul li") else { return [] }
        return items.compactMap { item in
            guard let time = try? item.select(".item-title").first()?.text().trimmingCharacters(in: .whitespaces),
                  let amountText = try? item.select(".item-after").first()?.text().trimmingCharacters(in: .whitespaces),
                  let amountMatch = amountText.range(of: "(-?[\\d.]+)元", options: .regularExpression) else { return nil }
            let amountStr = amountText[amountMatch].replacingOccurrences(of: "元", with: "")
            guard let amount = Double(amountStr) else { return nil }
            let typeText = (try? item.select(".item-subtitle").first()?.text().trimmingCharacters(in: .whitespaces)) ?? ""
            let payType = typeText.range(of: "类型:\\s*(.+)", options: .regularExpression).map { String(typeText[$0]).replacingOccurrences(of: "类型:", with: "").trimmingCharacters(in: .whitespaces) } ?? typeText
            return PaymentRecord(paymentTime: time, amount: amount, paymentType: payType)
        }
    }
}
